Dockerfile

```diff
@@ -0,0 +1,109 @@
+
+FROM rust:1.71.1 AS cargo-hack
+RUN cargo install -q cargo-hack
+
+FROM rust:1.71.1 AS cargo-kcov
+RUN cargo install -q cargo-kcov
+
+FROM rust:1.71.1 AS exa
+WORKDIR /app
+# Copy the source code into the image
+COPY . .
+# Build exa
+RUN cargo build
+
+# We use Ubuntu instead of Debian because the image comes with two-way
+# shared folder support by default.
+# This image is based from Ubuntu
+FROM rust:1.71.1 AS base
+
+# Install the dependencies needed for exa to build
+RUN apt-get update -qq
+RUN apt-get install -qq -o=Dpkg::Use-Pty=0 \
+    git gcc curl attr libgit2-dev zip \
+    fish zsh bash bash-completion
+
+# Install kcov for test coverage
+# This doesn’t run coverage over the xtests so it’s less useful for now
+COPY --from=cargo-kcov /usr/local/cargo/bin/cargo-kcov /usr/bin/cargo-kcov
+
+RUN apt-get install -y \
+    cmake g++ pkg-config \
+    libcurl4-openssl-dev libdw-dev binutils-dev libiberty-dev
+
+RUN ln -s $(which python3) /usr/bin/python
+RUN cargo kcov --print-install-kcov-sh | sh
+
+# Create a variety of misc scripts.
+
+RUN ln -sf /vagrant/devtools/dev-run-debug.sh /usr/bin/exa
+RUN ln -sf /vagrant/devtools/dev-run-release.sh /usr/bin/rexa
+
+RUN echo -e "#!/bin/sh\ncargo build --manifest-path /vagrant/Cargo.toml \\$@" > /usr/bin/build-exa
+RUN ln -sf /usr/bin/build-exa /usr/bin/b
+
+RUN echo -e "#!/bin/sh\ncargo test --manifest-path /vagrant/Cargo.toml \\$@ -- --quiet" > /usr/bin/test-exa
+RUN ln -sf /usr/bin/test-exa /usr/bin/t
+
+RUN echo -e "#!/bin/sh\n/vagrant/xtests/run.sh" > /usr/bin/run-xtests
+RUN ln -sf /usr/bin/run-xtests /usr/bin/x
+
+RUN echo -e "#!/bin/sh\nbuild-exa && test-exa && run-xtests" > /usr/bin/compile-exa
+RUN ln -sf /usr/bin/compile-exa /usr/bin/c
+
+ADD --chmod=+x devtools/dev-package-for-linux.sh /vagrant/devtools/dev-package-for-linux.sh
+RUN echo -e "#!/bin/sh\nbash /vagrant/devtools/dev-package-for-linux.sh \\$@" > /usr/bin/package-exa
+RUN echo -e "#!/bin/sh\ncat /etc/motd" > /usr/bin/halp
+
+# Configure the welcoming text that gets shown:
+
+# Capture the help text so it gets displayed first
+RUN rm -f /etc/update-motd.d/*
+ADD --chmod=+x devtools/dev-help.sh /vagrant/devtools/dev-help.sh
+RUN bash /vagrant/devtools/dev-help.sh > /etc/motd
+
+# Tell bash to execute a bunch of stuff when a session starts
+RUN echo "source /vagrant/devtools/dev-bash.sh" > ${HOME}/.bash_profile
+
+# Link the completion files so they’re “installed”:
+
+# bash
+RUN ln -s /vagrant/completions/bash/exa /etc/bash_completion.d/exa
+RUN ln -s /vagrant/completions/bash/exa /usr/share/bash-completion/completions/exa
+
+# zsh
+RUN ln -s /vagrant/completions/zsh/_exa /usr/share/zsh/vendor-completions/_exa
+
+# fish
+RUN ln -s /vagrant/completions/fish/exa.fish /usr/share/fish/completions/exa.fish
+
+
+# Copy the source code into the image
+COPY --chmod=+x . /vagrant/
+
+# Make sudo dummy replacement
+# This is needed for some tests that use sudo
+RUN echo -e '#!/bin/sh\n"$@"' > /usr/bin/sudo
+RUN chmod +x /usr/bin/sudo
```

Comment:
Really shouldn't be necessary to rely on sudo if the intent is to replace Vagrantfile with Dockerfile.

Reply:
I agree, I just don't want to touch all the internal files because a) I'm not too familiar with them and b) I'm not sure it'll behave the same as I couldn't run the Vagrant tests. I think we should go iteratively on this, by first making it compatible and afterwards refactoring the `sudo` stuff